### PR TITLE
[TM-1531] add job connection

### PR DESCRIPTION
**Pull request by egrojMonroy**
add the connection to pull the list of jobs

### CODE REVIEW

File: src/connections/DelayedJob.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    pathParams: { uuid }
+  };
+
+  return delayedJobsFindFetchFailed(variables)(store) != null;
+};
+const connectionIsLoaded = ({ delayedJobs, hasLoadFailed, isLoading }: DelayedJobsConnection) => {
+  return (delayedJobs != null && delayedJobs.length > 0) || hasLoadFailed || isLoading;
+};
+
+const delayedJobsConnection: Connection<DelayedJobsConnection, DelayedJobsProps> = {
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:35]: These props aren't used, no reason to include them.

File: src/connections/DelayedJob.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+const delayedJobsSelector = (store: ApiDataStore) => {
+  const delayedJobsMap = store.delayedJobs || {};
+  return Object.values(delayedJobsMap).map(resource => resource.attributes);
+};
+
+const delayedJobsLoadFailedSelector = (store: ApiDataStore, { uuid }: DelayedJobsProps) => {
+  const variables: DelayedJobsFindVariables = {
+    pathParams: { uuid }
+  };
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:27]: Since the API request doesn't use these path params, the load failed selector shouldn't either. The params to those two must always match for this feature to function.

File: src/connections/DelayedJob.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    const isLoaded = connectionIsLoaded(connection);
+    if (!isLoaded) {
+      listDelayedJobs();
+    }
+  },
+  isLoaded: connectionIsLoaded,
+  selector: selectorCache(
+    props => props.uuid,
+    props =>
+      createSelector(
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:46]: Since there are no props, the selector cache isn't needed - it's only needed when there are props to differentiate different selectors. The `selector` here can just receive `createSelector`

**Review by roguenet — COMMENTED** [review]
A quick review pass to hopefully nudge you closer to the right direction.

File: src/connections/DelayedJob.ts (hunk 2 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: understood and removed

File: src/connections/DelayedJob.ts (hunk 3 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: removed

File: src/connections/DelayedJob.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+// --- Delayed Jobs Connection ---
+type DelayedJobsConnection = {
+  delayedJobs?: DelayedJobDto[];
+  isLoading: boolean;
+  hasLoadFailed: boolean;
+};
+
+const delayedJobsSelector = (store: ApiDataStore) => {
+  const delayedJobsMap = store.delayedJobs || {};
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:23]: Please use ?? for defaults. I'd reformat this whole method:
```
const delayedJobsSelector = (store: ApiDataStore) => 
  Object.values(store.delayedJobs ?? {}).map(resource => resource.attributes);
```

File: src/connections/DelayedJob.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  return Object.values(delayedJobsMap).map(resource => resource.attributes);
+};
+
+const delayedJobsLoadFailedSelector = (store: ApiDataStore) => {
+  return listDelayedJobsFetchFailed(store) != null;
+};
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:29]: Single statement arrow functions should use the shortcut:
```
const delayedJobsLoadFailedSelector = (store: ApiDataStore) => listDelayedJobsFetchFailed(store) != null;
```

File: src/connections/DelayedJob.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+const connectionIsLoaded = ({ delayedJobs, hasLoadFailed, isLoading }: DelayedJobsConnection) => {
+  return (delayedJobs != null && delayedJobs.length > 0) || hasLoadFailed || isLoading;
+};
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:33]: Same here - please use the arrow function shortcut.

File: src/connections/DelayedJob.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+const delayedJobsConnection: Connection<DelayedJobsConnection> = {
+  load: connection => {
+    const isLoaded = connectionIsLoaded(connection);
+    if (!isLoaded) {
+      listDelayedJobs();
+    }
+  },
+  isLoaded: connectionIsLoaded,
+  selector: createSelector(
+    [delayedJobsSelector, store => delayedJobsLoadFailedSelector(store)],
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:44]: This second argument does not need to create a fresh function. This line can be
```
[delayedJobsSelector, delayedJobsLoadFailedSelector],
```

File: src/connections/DelayedJob.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    (delayedJobs, hasLoadFailed): DelayedJobsConnection => ({
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:45]: 🎨 It's OK to leave it, but I believe setting the return type here is not necessary (the type comes from the connection shape... if your return type didn't match DelayedJobsConnection, it would complain even if you don't specify it here):
```
(delayedJobs, hasLoadFailed) => ({
```

File: src/connections/DelayedJob.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      delayedJobs,
+      isLoading: delayedJobs == null && !hasLoadFailed,
+      hasLoadFailed
+    })
+  )
+};
+
+export const useDelayedJobs = () => {
+  const connection = connectionHook(delayedJobsConnection)();
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:54]: This should use `useConnection` directly.
```
const connection = useConnection(delayedJobsConnection);
```

File: src/connections/DelayedJob.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+  useEffect(() => {
+    const intervalId = setInterval(() => {
+      listDelayedJobs();
+    }, 1500);
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:59]: This will cause this interval to keep refreshing the delayed jobs for as long as the surrounding component is mounted. Is that intentional? It seems like clearing the interval when all the delayed jobs have completed would be good to keep from continuing to hit the jobs endpoint. I'm OK with letting this live to prod if that's complicated to wire up; it's just not ideal and should be cleaned up in a future ticket.

Please either decide to implement an eventual stop like that in this PR, or see that a tech debt ticket is created to clean up in a future release.

File: src/connections/DelayedJob.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    return () => {
+      clearInterval(intervalId);
+    };
+  }, []);
+
+  return connection;
+};
+
+export const triggerBulkUpdate = (jobs: DelayedJobData[]) => {
+  return bulkUpdateJobs({ body: { data: jobs } });
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:70]: I don't think there's any reason to return here. Functions that trigger v3 API calls should never be reliant on their return value; the data must always come from the redux store.

File: src/connections/DelayedJob.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+const bulkUpdateJobsSelector = (store: ApiDataStore) => {
+  const bulkUpdateState = store.delayedJobs || {};
+  return {
+    isLoading: bulkUpdateJobsIsFetching(store),
+    hasLoadFailed: bulkUpdateJobsFetchFailed(store) != null,
+    response: bulkUpdateState.response
+  };
+};
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:80]: This should use `??` for defaults, although I think this is an overall better format for this method:
```
const bulkUpdateJobsSelector = (store: ApiDataStore) => ({
  isLoading: bulkUpdateJobsIsFetching(store),
  hasLoadFailed: bulkUpdateJobsFetchFailed(store) != null,
  response: store.delayedJobs?.reponse
});
```

However, I'm also confused as to why you would expect a `response` object to be on `store.delayedJobs`. That shouldn't be possible with the current redux setup.

File: src/connections/DelayedJob.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+const bulkUpdateJobsConnection: Connection<DelayedJobsConnection, { jobs: DelayedJobData[] }> = {
+  load: async (connection, { jobs }) => {
+    const isLoaded = connectionBulkUpdateIsLoaded(connection);
+    if (!isLoaded) {
+      await bulkUpdateJobs({ body: { data: jobs } });
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:86]: connections should never have await / async mechanics. The data will be loaded into the redux store when it's been fetched, and then isLoaded / selector are called again.

File: src/connections/DelayedJob.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    }
+  },
+
+  isLoaded: state => !state.isLoading,
+
+  selector: createSelector([store => bulkUpdateJobsSelector(store)], ({ isLoading, hasLoadFailed, response }) => ({
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:92]: Please avoid creating extraneous intermediate functions:
```
selector: createSelector([bulkUpdateJobsSelector], ({
```

File: src/connections/DelayedJob.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    isLoading,
+    hasLoadFailed,
+    response
+  }))
+};
+
+const connectionBulkUpdateIsLoaded = ({ isLoading, hasLoadFailed }: { isLoading: boolean; hasLoadFailed: boolean }) => {
+  return !isLoading && !hasLoadFailed;
+};
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:101]: Please use the arrow function shortcut:
```
=> !isLoading && !hasLoadFailed;
```

File: src/store/apiSlice.ts [hunk 16]
```diff
@@ -45,6 +45,7 @@ export type Relationships = {
 };
 
 export type StoreResource<AttributeType> = {
+  id?: string;
```
Comment (roguenet) [inline src/store/apiSlice.ts:48]: Why was this added?

File: src/connections/DelayedJob.ts (hunk 4 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: changed

File: src/connections/DelayedJob.ts (hunk 5 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: changed

File: src/connections/DelayedJob.ts (hunk 6 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: changed

File: src/connections/DelayedJob.ts (hunk 7 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: understood and changed

File: src/connections/DelayedJob.ts (hunk 8 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: improved

File: src/connections/DelayedJob.ts (hunk 9 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: done

File: src/connections/DelayedJob.ts (hunk 11 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: understood, and changed

File: src/connections/DelayedJob.ts (hunk 13 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: removed

File: src/connections/DelayedJob.ts (hunk 14 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: changed

File: src/connections/DelayedJob.ts (hunk 15 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: changed

File: src/store/apiSlice.ts (hunk 16 above)
Comment (egrojMonroy) [reply src/store/apiSlice.ts]: this should have not reached git. It was from an understanding of the store. Removed

File: src/connections/DelayedJob.ts (hunk 12 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: the response object wasnt needed, and everything keeps working.

File: src/connections/DelayedJob.ts (hunk 10 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts:52]: Yes indeed, it is supposed to be attentive for any job created on any session. 
Nonetheless I think it might not be good. 
So for now the debt ticket will be needed, because we would have to think on the user cases.

File: src/connections/DelayedJob.ts [hunk 17]
```diff
@@ -66,38 +59,32 @@ export const useDelayedJobs = () => {
...
     }
   },
 
   isLoaded: state => !state.isLoading,
 
-  selector: createSelector([store => bulkUpdateJobsSelector(store)], ({ isLoading, hasLoadFailed, response }) => ({
+  selector: createSelector(bulkUpdateJobsSelector, ({ isLoading, hasLoadFailed, response }) => ({
     isLoading,
     hasLoadFailed,
     response
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:83]: This seems like it should be throwing an error in the TS compiler... Maybe I'm missing something, but this shape does not seem to match `DelayedJobsConnection`. 

A couple of thoughts here:
* Typically, each connection will have its own shape. That's not required, but it's normal
* I've found in the past that properties like `isLoading`, `hasLoadFailed` and `response`, while seeming convenient when looking at a single connection, are actually a disaster down the road. The reason being that if we follow such generic conventions for individual connections, then when a component ends up using 2 or more connections, it gets really awkward to juggle all the `isLoading`s that each connection returns. Instead, I think it's better to be super specific with the properties of the connection shape. So something more like `bulkUpdateJobsIsLoading` and `bulkUpdateJobsHasFailed`.

As for `response` - this appears to be just returning all the delayed jobs like the other connection does. Are these used in the same component? Perhaps a better approach would be to combine them into a single connection. I think of connections as being responsible for providing a given set of data from the redux cache, and a set of functions for operating on that data. It would be totally reasonable for this `bulkUpdateJobs` call to be triggered from a function that's included on the other connection. Or, the bulkUpdate function could be simply exposed by this file, and then we simply rely on the original connection to provide the `delayedJobs` from the store. One of the primary wins of this system is that a connection returns data that got cached in the store _regardless of where it came from_. So if one call does the initial fetch, and another call modifies that data, as long as the data ends up back in the store under the same ID, the component that relies on it will rerender when it gets updated.

Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: Finally joined both connections. 
Tested and is working.

File: src/components/elements/Notification/FloatNotification.tsx [hunk 18]
```diff
@@ -38,36 +83,81 @@ const FloatNotification = ({ data }: FloatNotificationProps) => {
...
+                                  : ((item.processedContent ?? 0) / (item.totalContent ?? 1)) * 100
+                              }
+                              className="h-2 bg-success-40"
+                              color="success-600"
+                            />
+                          )}
+                          <Text variant="text-12-semibold" className="text-black">
+                            {item.name === "Polygon Upload"
+                              ? item.status === "succeeded"
+                                ? "Done!"
```
Comment (roguenet) [inline src/components/elements/Notification/FloatNotification.tsx:138]: Please make sure any text that is displayed gets wrapped in `t()`

File: src/connections/DelayedJob.ts [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+const delayedJobsSelector = (store: ApiDataStore) =>
+  Object.values(store.delayedJobs ?? {}).map(resource => resource.attributes);
+
+const bulkUpdateJobsSelector = (store: ApiDataStore) => ({
+  bulkUpdateJobsIsLoading: bulkUpdateJobsIsFetching(store),
+  bulkUpdateJobsHasFailed: bulkUpdateJobsFetchFailed(store) != null,
+  updatedJobsResponse: Object.values(store.delayedJobs ?? {}).map(resource => resource.attributes as DelayedJobDto)
+});
```
Comment (roguenet) [inline src/connections/DelayedJob.ts:27]: These two selectors are problematic because they're inputs to `createSelector`, and create a new object every time they're called. If their inputs don't change, they should return the same object every time to gain the true benefit of reselect. 

A better approach is for the delayedJobsSelector to be simpler: 
```
const delayedJobsSelector = (store: ApiDataStore) => store.delayedJobs
```

and to remove this bulk update selecot, instead relying directly on the isFetching / isFailed selectors in the `createSelector` call. So it becomes something more like:
```
const combinedSelector = createSelector([
  delayedJobsSelector, bulkUpdateJobsIsFetching, bulkUpdateJobsFetchFailed
], (delayedJobs, bulkUpdateJobsIsLoading, bulkUpdateJobsFailure) => ({
  delayedJobs: Object.values(delayedJobs ?? {}).map(resource => resource.attributes),
  ... // other members of the connection
}));
```

It's important for the members of the first argument array to `createSelector` to only return a unique object result if the inputs have changed, whenever possible.

File: src/components/elements/Notification/FloatNotification.tsx (hunk 18 above)
Comment (egrojMonroy) [reply src/components/elements/Notification/FloatNotification.tsx]: added to every place but names

File: src/connections/DelayedJob.ts (hunk 19 above)
Comment (egrojMonroy) [reply src/connections/DelayedJob.ts]: changed it and is working still.

**Review by roguenet — APPROVED** [review]
